File: patterns/c#/TopKElements.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class TopKElements {
    
    // K Largest Elements using Sorting
    public int[] KLargestElementsSortingApproach(int[] nums, int k) {
        Array.Sort(nums, (a, b) => b.CompareTo(a));
        return nums.Take(k).ToArray();
    }

    // K Largest Elements using Max Heap
    public int[] KLargestElementsMaxHeapApproach(int[] nums, int k) {
        PriorityQueue<int, int> maxHeap = new PriorityQueue<int, int>(Comparer<int>.Create((a, b) => b - a));
        foreach (var num in nums) {
            maxHeap.Enqueue(num, num);
        }
        var result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = maxHeap.Dequeue();
        }
        return result;
    }

    // K Largest Elements using Min Heap
    public int[] KLargestElementsMinHeapApproach(int[] nums, int k) {
        PriorityQueue<int, int> minHeap = new PriorityQueue<int, int>();
        for (int i = 0; i < k; i++) {
            minHeap.Enqueue(nums[i], nums[i]);
        }
        for (int i = k; i < nums.Length; i++) {
            minHeap.Enqueue(nums[i], nums[i]);
            if (minHeap.Count > k) {
                minHeap.Dequeue();
            }
        }
        var result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.Dequeue();
        }
        return result;
    }

    // Top K Frequent Elements using Sorting
    public int[] TopKFrequentElementsSortingApproach(int[] nums, int k) {
        var frequencyMap = new Dictionary<int, int>();
        foreach (var num in nums) {
            if (!frequencyMap.ContainsKey(num)) {
                frequencyMap[num] = 0;
            }
            frequencyMap[num]++;
        }

        var sortedEntries = frequencyMap.OrderByDescending(e => e.Value).Take(k).Select(e => e.Key).ToArray();
        return sortedEntries;
    }

    // Top K Frequent Elements using Min Heap
    public int[] TopKFrequentElementsMinHeapApproach(int[] nums, int k) {
        var frequencyMap = new Dictionary<int, int>();
        foreach (var num in nums) {
            if (!frequencyMap.ContainsKey(num)) {
                frequencyMap[num] = 0;
            }
            frequencyMap[num]++;
        }

        var minHeap = new PriorityQueue<int, int>(Comparer<int>.Create((a, b) => a.CompareTo(b)));
        foreach (var entry in frequencyMap) {
            minHeap.Enqueue(entry.Key, entry.Value);
            if (minHeap.Count > k) {
                minHeap.Dequeue();
            }
        }

        var result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.Dequeue();
        }
        return result;
    }

    // K Closest Points to Origin using Max Heap
    private int GetDistance(int[] point) {
        return point[0] * point[0] + point[1] * point[1];
    }

    public int[][] KClosestPointsToOriginMaxHeapApproach(int[][] points, int k) {
        PriorityQueue<int[], int> maxHeap = new PriorityQueue<int[], int>(Comparer<int>.Create((a, b) => b - a));
        foreach (var point in points) {
            maxHeap.Enqueue(point, GetDistance(point));
            if (maxHeap.Count > k) {
                maxHeap.Dequeue();
            }
        }
        var result = new int[k][];
        for (int i = 0; i < k; i++) {
            result[i] = maxHeap.Dequeue();
        }
        return result;
    }
}

File: patterns/typescript/top_k_elements.ts
class TopKElements {
    
    // K Largest Elements using Sorting
    kLargestElementsSortingApproach(nums: number[], k: number): number[] {
        nums.sort((a, b) => b - a);
        return nums.slice(0, k);
    }

    // K Largest Elements using Max Heap
    kLargestElementsMaxHeapApproach(nums: number[], k: number): number[] {
        const maxHeap = new MaxPriorityQueue({ priority: (x: number) => x });
        for (const num of nums) {
            maxHeap.enqueue(num);
        }
        const result: number[] = [];
        for (let i = 0; i < k; i++) {
            result.push(maxHeap.dequeue().element);
        }
        return result;
    }

    // K Largest Elements using Min Heap
    kLargestElementsMinHeapApproach(nums: number[], k: number): number[] {
        const minHeap = new MinPriorityQueue({ priority: (x: number) => x });
        for (let i = 0; i < k; i++) {
            minHeap.enqueue(nums[i]);
        }
        for (let i = k; i < nums.length; i++) {
            minHeap.enqueue(nums[i]);
            if (minHeap.size() > k) {
                minHeap.dequeue();
            }
        }
        const result: number[] = [];
        for (let i = 0; i < k; i++) {
            result.push(minHeap.dequeue().element);
        }
        return result;
    }

    // Top K Frequent Elements using Sorting
    topKFrequentElementsSortingApproach(nums: number[], k: number): number[] {
        const frequencyMap = new Map<number, number>();
        nums.forEach(num => frequencyMap.set(num, (frequencyMap.get(num) || 0) + 1));
        return Array.from(frequencyMap)
            .sort((a, b) => b[1] - a[1])
            .slice(0, k)
            .map(entry => entry[0]);
    }

    // Top K Frequent Elements using Min Heap
    topKFrequentElementsMinHeapApproach(nums: number[], k: number): number[] {
        const frequencyMap = new Map<number, number>();
        nums.forEach(num => frequencyMap.set(num, (frequencyMap.get(num) || 0) + 1));
        const minHeap = new MinPriorityQueue({ priority: (x: [number, number]) => x[1] });
        frequencyMap.forEach((value, key) => {
            minHeap.enqueue([key, value]);
            if (minHeap.size() > k) {
                minHeap.dequeue();
            }
        });
        const result: number[] = [];
        for (let i = 0; i < k; i++) {
            result.push(minHeap.dequeue().element[0]);
        }
        return result;
    }

    // K Closest Points to Origin using Max Heap
    getDistance(point: number[]): number {
        return point[0] ** 2 + point[1] ** 2;
    }

    kClosestPointsToOriginMaxHeapApproach(points: number[][], k: number): number[][] {
        const maxHeap = new MaxPriorityQueue({ priority: (point: number[]) => -this.getDistance(point) });
        points.forEach(point => {
            maxHeap.enqueue(point);
            if (maxHeap.size() > k) {
                maxHeap.dequeue();
            }
        });
        const result: number[][] = [];
        for (let i = 0; i < k; i++) {
            result.push(maxHeap.dequeue().element);
        }
        return result;
    }
}